### ISSUE TITLE
implement dns cache

### ISSUE THREAD

**Issue by lublak** [algesten/ureq#401]
Currently ureq queries the dns server with every query, which is not exactly optimal, especially if you have a slow dns server.
A library that has it implemented by default would be trust-dns-resolver.
This option could be optional.

**Comment by algesten** [comment]
Hi! Welcome to ureq!

There are a few reasons why I think DNS caching isn't necessarily something that belongs in ureq.

1. ureq in itself doesn't actually do the DNS querying. The call is provided by Rust's built in [`ToSocketAddrs `](https://github.com/algesten/ureq/blob/main/src/resolve.rs#L16), which in turn relies on the host as a resolver. It's quite common for the host OS to provide DNS caching, which would then implicitly be used by ureq.

2. ureq's DNS resolving is also pluggable, which means that if someone really doesn't want to rely on the OS, it's quite easy to write a bespoke strategy for caching. Here is an example of how to use just ipv6 addresses, it can be serve as a starting point for caching, or using trust-dns-resolver. https://github.com/algesten/ureq/blob/main/examples/ipv6.rs

3. trust-dns-resolver would introduce quite a few dependencies, and we try to keep ureq lean.

Overall, I'm quite happy with the current state of ureq's DNS resolving mechanism. The one thing that I would like to have a better solution for, is that `ToSocketAddrs ` doesn't provide a mechanism to time out the DNS query. This currently forces us into some ugly hacks to get around that. trust-dns-resolver, although it is a really nice crate, would however be a a sledge hammer to crack that small nut.

**Comment by lublak** [comment]
Hi @algesten 😄 

1. I have only sensed one kind of delay with ureq as opposed to reqwest. I deactivated the function of trust-dns-resolver at reqwest, which reqwest has as an option, and the same delay occurred. There may be other things that play into it.
2. Yes i found it out yesterday and build a TrustDNSResolver. But its also slow with default config. I have not yet tried the system conf which reqwest uses in trust dns.
3. yes ... sadly
I have been looking for an alternative because Reqwest has taken my 200kb application to 4000kb.
Then I found ureq which ended up at 600kb in total. It is a simple application that is also continuously synchronised over the network. Therefore, the size is also important.
Trust-Dns is not much better. from 200kb to 2000kb.

I am also relatively new to the Rust world. I'm still excited about it though, as I've never really liked c or c++. Rust just feels more right. The only thing missing here are the one or other libraries that work with safe code and you don't have to write unsafe rust.
And yes trust-dns-resolver might not be the best alternative after all. I'll see if I can find something smaller that only implements the system_conf of trust-dns and without tokio dependency.

Thanks for your answer
